bin/kano-updater-recovery

```diff
@@ -0,0 +1,351 @@
+    """
+    Parses the return code from the updater terminating to the language of the
+    UFx codes.
+
+    Args:
+        rc (int): Return code from the updater's termination
+
+    Returns:
+        int: Return code corresponding to the updater's finished state (Member
+        of :class:`RC` class).
+    """
+
+    from kano_updater.return_codes import RC as UpdaterRC
+
+    if rc == UpdaterRC.SUCCESS:
+        reset_wifi_retries()
+        return rc
+
+    elif rc in [UpdaterRC.NO_NETWORK, UpdaterRC.CANNOT_REACH_KANO]:
+        if get_wifi_retries() > PERMITTED_WIFI_RETRIES:
+            return RC.UF4
+        else:
+            increment_wifi_retries()
+            return RC.UF3
+
+    elif rc == UpdaterRC.HANGED_INDEFINITELY:
+        return RC.UF5
+
+    else:
+        return RC.UF2
+
+
+def reboot(delay=0):
+    """Perform a system reboot.
+
+    Args:
+        delay (int): Time in seconds to delay the operation (same thread).
+    """
+
+    if delay:
+        import time
+        time.sleep(delay)
+
+    import os
+    os.system('systemctl reboot')
+
+
+def prepare_for_hard_shutdown():
+    """Configure the system to handle a hard poweroff gracefully."""
+
+    pass  # TODO: Insert black magic here.
+
+
+def infinite_loop():
+    """Start an infinite loop within the main thread.
+
+    Note:
+        This function is designed to be the last thing you call. It will
+        block indefinitely.
+    """
+
+    import time
+    while True:
+        time.sleep(1 * HOUR)
+
+
+def main():
+    """Main recovery program."""
+
+    from kano_updater.status import UpdaterStatus
+    if not UpdaterStatus.get_instance().is_recovery_needed():
+        return RC.SUCCESS
+
+    # Check if any UFx is being shown, e.g. kano-os-recovery had a timeout.
+    if is_err_being_shown():
+        prepare_for_hard_shutdown()
+        infinite_loop()
+        return RC.SUCCESS
+
+    show_recovery_splash()
+
+    proc = run_updater()
+    rc = 0
+
+    import time
+    start_time = time.time()
+    duration = 0
+
+    # Poll the status of the Updater process and implement an overall timeout
+    # to prevent this from hanging and/or taking too long.
+    while True:
+        updater_rc = proc.poll()
+
+        if updater_rc is not None:
+            rc = parse_updater_rc(updater_rc)
+
+            if rc == RC.SUCCESS:
+                show_success_splash()
+                reboot(delay=10)
+                return rc
+            else:
+                show_err(rc)
+                break
+
+        now = time.time()
+        duration = now - start_time
+
+        if duration > SYSTEM_PROGRESS_TIMEOUT:
+            stop_updater(proc)
+            rc = show_err(RC.UF6)
+            break
+
+        time.sleep(1)
+
+    # If the script has gotten to this point, a UFx is being shown. It's
+    # expected that the recovery shows the error indefinitely until the user
+    # unplugs to power off.
+    prepare_for_hard_shutdown()
+    infinite_loop()
```

Comment:
If we've popped up one of the error codes using the `recovery-err-splash@.service` unit, the message won't timeout so why do we need to keep this alive with the infinite loop?

Reply:
This will be because this script is started from `kano-ui-autostart`. As such, if the script terminates, the other will continue to go to Dashboard. Maybe when we convert that one into a `.target` and move everything into `.services` thereafter..